server/player/player.go

```diff
@@ -557,9 +557,38 @@ func (p *Player) Hurt(dmg float64, source damage.Source) (float64, bool) {
 		p.Exhaust(0.1)
 
 		damageToArmour := int(math.Max(math.Floor(dmg/4), 1))
+		var damageToAttacker int
+		thornsArmour := make([]item.Stack, 0, 4)
 		for slot, it := range p.armour.Slots() {
 			if _, ok := it.Item().(item.Durable); ok {
-				_ = p.armour.Inventory().SetItem(slot, p.damageItem(it, damageToArmour))
+				var thornsDamage int
+				if e, ok := it.Enchantment(enchantment.Thorns{}); ok && rand.Float64() < e.Level()*0.15 {
+					thornsArmour = append(thornsArmour, it)
+					thornsDamage = 1
+					if e.Level() > 10 {
+						damageToAttacker += e.Level() - 10
+					} else {
+						damageToAttacker += 1 + rand.Intn(4)
+					}
+				}
+				_ = p.armour.Inventory().SetItem(slot, p.damageItem(it, damageToArmour+thornsDamage))
+			}
+		}
+		if length := len(thornsArmour); length > 0 {
+			slot := rand.Intn(length)
+			it := thornsArmour[slot]
```

Comment:
```suggestion
			slot := rand.Intn(length)
			it := thornsArmour[slot]
```

This will not set to the correct slot unless all four armor pieces proc Thorns.